support local creds probing

yugangw-msft opened the pull request:
For #118. 
The probe sequence:
1. Manager service identity
   only "system assigned identity" is probed, as user assigned identity would require extra arguments, and under that context users are better off using strong type credentials 
2. Environment variable `AZURE_CONNECTION_STRING` with sdk auth file content in json, which is proposed in the design review meeting in June.
    e.g. `{AZURE_CLIENT_ID='...', AZURE_CLIENT_SECRET='...', AZURE_TENANT_ID='...', AZURE_CLOUD_NAME='...'}`
    `AZURE_CLOUD_NAME` is optional. We default to public cloud
3. Individual environment variables for a service principal
    Like 2, rather using individual environment variables
4. Azure Authentication File (originating from java sdk)
5. Azure CLI credentials

Subscription id will be optional as the chainer will find them
Per suggestion, the example code includes both management plane and data plane client codes

yugangw-msft commented:
//cc: @lmazuel  @jianghaolu

yugangw-msft commented:
cc: @jianghaolu  @lmazuel  @jhendrixMSFT @shahabhijeet @joshgav

lmazuel commented on msrestazure/azure_local_creds_prober.py:
```diff
@@ -0,0 +1,142 @@
+            program_files_folder = os.environ.get('ProgramFiles(x86)') or os.environ.get('ProgramFiles')
+            probing_paths = [os.path.join(program_files_folder, 'Microsoft SDKs', 'Azure', 'CLI2', 'wbin', 'az.cmd')]
+        else:
+            probing_paths = ['/usr/bin/az', '/usr/local/bin/az']
```
This won't work if CLI was installed by pip. You could subprocess "which az", or us "shutil.which('az')" starting Python 3
https://docs.python.org/3/library/shutil.html?highlight=.which#shutil.which

johanste replied:
Can we use the presence of a ${HOME}/.azure file as the hint to try the CLI creds?

johanste replied:
...or, alternatively, use the ${PATH} environment variable...

lmazuel commented on msrestazure/azure_local_creds_prober.py:
```diff
@@ -0,0 +1,142 @@
+
+    def signed_session(self, session=None):
+        # Token cache is handled by the VM extension, call each time to avoid expiration
+        self.set_token()
```
Should, the CLI be called each request to refresh the token?

yugangw-msft replied:
I assume you meant sub-shelling is kind of slow? Yes, I can optimize by leveraging the `expiresOn` information returned from the `get-access-token` command

lmazuel commented on tests/test_local_creds_prober.py:
```diff
@@ -0,0 +1,34 @@
+from msrestazure.azure_local_creds_prober import AzureLocalCredentialProber
+
+prober = AzureLocalCredentialProber()
+client = StorageManagementClient(prober, prober.subscription_id)
```
I would prefer to use the factory style I have right now:
```python
from azure.common.client_factory import get_client_from_cli_profile
from azure.mgmt.compute import ComputeManagementClient

client = get_client_from_cli_profile(ComputeManagementClient)
```

but something more generic like:

```python
from azure.common.client_factory import get_client
from azure.mgmt.compute import ComputeManagementClient

client = get_client(ComputeManagementClient)
```

johanste commented on msrestazure/azure_local_creds_prober.py:
```diff
@@ -0,0 +1,142 @@
+class AzureLocalCredentialProber(object):
+    '''
+    Probing logics:
+    1. Managed service identity
```
The progression should be from more specific to less specific. This is the way I'd define the specificity of the different sources of information:

```
Environment variables > local file > local endpoint available. 
```
Following this, does this: 2, 3, 1, 4 make sense?

yugangw-msft replied:
I started with the most secure option, then follow up with SDK home made ones. But you have a point, so I will adjust. Also it is unlikely people would set env variable in a VM with identity

joshgav commented:
Explicit environment variables should come before managed identity like @johanste says cause we should prefer explicit intent. That's what the specs say too :).

Rather than a global "Prober", could each credentials type include a static `Probe()` method which returns a boolean? That would grant more flexibility to the provider and allow us to add new providers and probes in the future.

Can we document the names of keys in the connection string and names of environment variables? We should use the same names in all SDKs.

joshgav commented:
We will need to get tokens for resources other than AzureRM, e.g. Storage and Key Vault. Can the constructor be parameterized to take a resource URL and default to "https://management.azure.com"?

yugangw-msft commented:
> We will need to get tokens for resources other than AzureRM, e.g. Storage and Key Vault

Each specific SDK has the resource uri defined inside, e.g. [go-sdk](https://github.com/Azure/azure-sdk-for-go/blob/master/services/keyvault/mgmt/2015-06-01/keyvault/client.go#L29) , so this is not a  concern.
Since we are at it, let us see how to get sovereign clouds work. In Python's autorest runtime, those endpoints are pre-defined as [a set of instances](https://github.com/Azure/msrestazure-for-python/blob/master/msrestazure/azure_cloud.py#L131).  I believe Java does too (cc: @jianghaolu).  If other language SDKs have the similar supports, then the probing can default to public azure while client app can override with an enum.

yugangw-msft commented:
Hi folks, i have addressed all feedback, please take another look.
The sample code wrapped in the unit test works for both storage management plane and data plane. Note the new storage data plan involves a highly unique requirement on client applications on how to supply the access token.  I will cross check with my CLI colleague for whether or not I am doing the most optimal way. 
The method now exposes the `resource` keyword arguments with the defaulting value on arm.

lmazuel commented on msrestazure/azure_local_creds_prober.py:
```diff
@@ -0,0 +1,280 @@
+    def probe(self):
+        creds = None
+        if self.enabled and os.environ.get('AZURE_CONNECTION_STRING'):
+            auth_info = json.loads(os.environ.get('AZURE_CONNECTION_STRING'))
```
Is the connection string JSON valid? I thought it was something like:
AZURE_CONNECTION_STRING="client_id=1234;client_secret=2345;tenant_id=3456;subscription_id=4567;"

lmazuel commented on msrestazure/azure_local_creds_prober.py:
```diff
@@ -0,0 +1,280 @@
+        return None
+
+    def probe_subscription(self, creds):
+        return json.loads(os.environ.get('AZURE_CONNECTION_STRING'))['subscriptionId']
```
Same question about JSON, and I thought too it should be `subscription_id`

lmazuel commented on msrestazure/azure_local_creds_prober.py:
```diff
@@ -0,0 +1,280 @@
+            else:
+                cloud_environment = AZURE_PUBLIC_CLOUD
+            resource = self.resource or cloud_environment.endpoints.management
+            creds = ServicePrincipalCredentials(client_id=auth_info['clientId'],
```
client_id?

lmazuel commented on msrestazure/azure_local_creds_prober.py:
```diff
@@ -0,0 +1,280 @@
+        a. app service
+        b. virtual machine
+    4. Azure CLI, through "az account get-access-token"
+    '''
```
Add the authentication file:
https://docs.microsoft.com/en-us/python/azure/python-sdk-azure-authenticate?view=azure-python#mgmt-auth-file

lmazuel commented on msrestazure/azure_local_creds_prober.py:
```diff
@@ -0,0 +1,280 @@
+
+def get_creds_through_local_probing(**kwargs):
+    '''
+    Probing logics:
```
Should we have the concept of critical failure? If we recognize some env variables and something is wrong, is this enough to fail completly and not fallback?
@johanste @joshgav

johanste replied:
Yes.

lmazuel commented:
Will be done in azure-identity but not msrestazure
